File: wrappers/java/src/main/java/com/turbotoken/Encoding.java
package com.turbotoken;

import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.OptionalInt;

/**
 * A loaded BPE encoding. Thread-safe after construction.
 *
 * Use {@link TurboToken#getEncoding(String)} or
 * {@link TurboToken#getEncodingForModel(String)} to obtain instances.
 */
public final class Encoding {

    private final byte[] rankPayload;
    private final Registry.EncodingSpec spec;

    Encoding(byte[] rankPayload, Registry.EncodingSpec spec) {
        this.rankPayload = rankPayload;
        this.spec = spec;
    }

    /** Returns the encoding spec. */
    public Registry.EncodingSpec getSpec() {
        return spec;
    }

    /** Returns the encoding name. */
    public String getName() {
        return spec.getName();
    }

    /* ── Core operations ─────────────────────────────────────────────── */

    /**
     * Encodes text into BPE token IDs.
     */
    public int[] encode(String text) {
        byte[] textBytes = text.getBytes(StandardCharsets.UTF_8);
        return NativeBridge.encodeBpe(rankPayload, textBytes);
    }

    /**
     * Decodes BPE token IDs back to a string.
     */
    public String decode(int[] tokens) {
        byte[] bytes = NativeBridge.decodeBpe(rankPayload, tokens);
        return new String(bytes, StandardCharsets.UTF_8);
    }

    /**
     * Counts the number of tokens in the given text without allocating the token array.
     */
    public int count(String text) {
        byte[] textBytes = text.getBytes(StandardCharsets.UTF_8);
        long c = NativeBridge.countBpe(rankPayload, textBytes);
        return (int) c;
    }

    /**
     * Alias for {@link #count(String)}.
     */
    public int countTokens(String text) {
        return count(text);
    }

    /**
     * Checks if the text is within the given token limit.
     * Returns the token count if within the limit, or empty if exceeded.
     */
    public OptionalInt isWithinTokenLimit(String text, int limit) {
        byte[] textBytes = text.getBytes(StandardCharsets.UTF_8);
        long result = NativeBridge.isWithinTokenLimit(rankPayload, textBytes, limit);
        if (result == -2) {
            return OptionalInt.empty();
        }
        return OptionalInt.of((int) result);
    }

    /* ── Chat operations ─────────────────────────────────────────────── */

    /**
     * Encodes a list of chat messages into token IDs.
     */
    public int[] encodeChat(List<ChatTemplate.ChatMessage> messages, ChatTemplate.ChatOptions options) {
        String formatted = ChatTemplate.formatMessages(messages, options);
        return encode(formatted);
    }

    /**
     * Counts tokens in a list of chat messages.
     */
    public int countChat(List<ChatTemplate.ChatMessage> messages, ChatTemplate.ChatOptions options) {
        String formatted = ChatTemplate.formatMessages(messages, options);
        return count(formatted);
    }

    /**
     * Checks if a chat conversation is within the token limit.
     */
    public OptionalInt isChatWithinTokenLimit(List<ChatTemplate.ChatMessage> messages,
                                               ChatTemplate.ChatOptions options, int limit) {
        String formatted = ChatTemplate.formatMessages(messages, options);
        return isWithinTokenLimit(formatted, limit);
    }

    /* ── File operations ─────────────────────────────────────────────── */

    /**
     * Encodes the contents of a file into BPE token IDs.
     */
    public int[] encodeFilePath(String filePath) {
        return NativeBridge.encodeBpeFile(rankPayload, filePath);
    }

    /**
     * Counts tokens in a file without reading it into Java memory.
     */
    public int countFilePath(String filePath) {
        long c = NativeBridge.countBpeFile(rankPayload, filePath);
        return (int) c;
    }

    /**
     * Checks if a file's contents are within the token limit.
     */
    public OptionalInt isFilePathWithinTokenLimit(String filePath, int limit) {
        // File-level limit check goes through encode and checks length
        int[] tokens = encodeFilePath(filePath);
        if (tokens.length <= limit) {
            return OptionalInt.of(tokens.length);
        }
        return OptionalInt.empty();
    }
}

File: wrappers/scala/src/main/scala/com/turbotoken/Encoding.scala
package com.turbotoken

import scala.jdk.CollectionConverters._

/** A loaded BPE encoding. Thread-safe after construction.
  *
  * Wraps the Java com.turbotoken.Encoding, providing Scala-idiomatic collection types.
  *
  * Use [[TurboToken.getEncoding]] or [[TurboToken.getEncodingForModel]] to obtain instances.
  */
class Encoding private[turbotoken] (private val javaEncoding: com.turbotoken.Encoding) {

  /** Returns the encoding name. */
  def name: String = javaEncoding.getName()

  /* ── Core operations ──────────────────────────────────────────── */

  /** Encodes text into BPE token IDs. */
  def encode(text: String): Seq[Int] =
    javaEncoding.encode(text).toSeq

  /** Decodes BPE token IDs back to a string. */
  def decode(tokens: Seq[Int]): String =
    javaEncoding.decode(tokens.toArray)

  /** Counts the number of tokens in the given text without allocating the token array. */
  def count(text: String): Int =
    javaEncoding.count(text)

  /** Alias for [[count]]. */
  def countTokens(text: String): Int = count(text)

  /** Checks if the text is within the given token limit.
    * Returns Some(tokenCount) if within the limit, or None if exceeded.
    */
  def isWithinTokenLimit(text: String, limit: Int): Option[Int] = {
    val result = javaEncoding.isWithinTokenLimit(text, limit)
    if (result.isPresent) Some(result.getAsInt) else None
  }

  /* ── Chat operations ──────────────────────────────────────────── */

  /** Encodes a list of chat messages into token IDs. */
  def encodeChat(messages: Seq[ChatMessage], options: ChatOptions = ChatOptions()): Seq[Int] = {
    val formatted = Chat.formatMessages(messages, options)
    encode(formatted)
  }

  /** Counts tokens in a list of chat messages. */
  def countChat(messages: Seq[ChatMessage], options: ChatOptions = ChatOptions()): Int = {
    val formatted = Chat.formatMessages(messages, options)
    count(formatted)
  }

  /** Checks if a chat conversation is within the token limit. */
  def isChatWithinTokenLimit(messages: Seq[ChatMessage], limit: Int, options: ChatOptions = ChatOptions()): Option[Int] = {
    val formatted = Chat.formatMessages(messages, options)
    isWithinTokenLimit(formatted, limit)
  }

  /* ── File operations ──────────────────────────────────────────── */

  /** Encodes the contents of a file into BPE token IDs. */
  def encodeFilePath(path: String): Seq[Int] =
    javaEncoding.encodeFilePath(path).toSeq

  /** Counts tokens in a file without reading it into Scala/Java memory. */
  def countFilePath(path: String): Int =
    javaEncoding.countFilePath(path)

  /** Checks if a file's contents are within the token limit. */
  def isFilePathWithinTokenLimit(path: String, limit: Int): Option[Int] = {
    val result = javaEncoding.isFilePathWithinTokenLimit(path, limit)
    if (result.isPresent) Some(result.getAsInt) else None
  }
}
